File: 310.minimum-height-trees.py
#
# @lc app=leetcode id=310 lang=python3
#
# [310] Minimum Height Trees
#

# Time:  O(n)
# Space: O(n)

from collections import defaultdict
from typing import List


# @lc code=start
class Solution:
    def findMinHeightTrees(self, n: int, edges: List[List[int]]) -> List[int]:
        if n == 1:
            return [0]
        neighbors = defaultdict(set)
        for a, b in edges:
            neighbors[a].add(b)
            neighbors[b].add(a)

        res, prev_leaves = set(), []
        for i in range(n):
            res.add(i)
            if len(neighbors[i]) == 1:
                prev_leaves.append(i)
        while len(res) > 2:
            curr_leaves = []
            for src in prev_leaves:
                res.remove(src)
                for dst in neighbors[src]:
                    if dst in res:
                        neighbors[dst].remove(src)
                        if len(neighbors[dst]) == 1:
                            curr_leaves.append(dst)
            prev_leaves = curr_leaves
        return list(res)


# @lc code=end

File: 310.minimum-height-trees.js
/*
 * @lc app=leetcode id=310 lang=javascript
 *
 * [310] Minimum Height Trees
 */

// @lc code=start
/**
 * @param {number} n
 * @param {number[][]} edges
 * @return {number[]}
 */
var findMinHeightTrees = function (n, edges) {
  if (n === 1) {
    return [0];
  }
  const neighbors = {};
  for (const edge of edges) {
    if (!neighbors[edge[0]]) {
      neighbors[edge[0]] = new Set();
    }
    if (!neighbors[edge[1]]) {
      neighbors[edge[1]] = new Set();
    }
    neighbors[edge[0]].add(edge[1]);
    neighbors[edge[1]].add(edge[0]);
  }

  const res = new Set();
  let prevLeaves = [];
  for (let i = 0; i < n; ++i) {
    res.add(i);
    if (neighbors[i].size === 1) {
      prevLeaves.push(i);
    }
  }
  while (res.size > 2) {
    const currLeaves = [];
    for (const src of prevLeaves) {
      res.delete(src);
      for (const dst of neighbors[src]) {
        if (res.has(dst)) {
          neighbors[dst].delete(src);
          if (neighbors[dst].size === 1) {
            currLeaves.push(dst);
          }
        }
      }
    }
    prevLeaves = currLeaves;
  }
  return Array.from(res);
};
// @lc code=end
